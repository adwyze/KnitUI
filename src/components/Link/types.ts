import { ReactNode } from "react"
import { BaseComponentProps } from "../../common/types"

export interface ILinkProps extends BaseComponentProps {
  href?: string
  /** whether to show underline or not, default to true */
  underline?: boolean
  children: ReactNode
  /** Disabled link, default to false */
  disabled?: boolean
  /** Opens link in a new tab, default to false */
  newTab?: boolean
}
