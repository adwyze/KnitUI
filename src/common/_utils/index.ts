import { ThemedStyledFunction } from "styled-components"
import chroma from "chroma-js"
import { ColorPreset, CustomColor } from "../types"

export const insertIf = (
  obj: any = {},
  condition: boolean,
  isArray: boolean = false
) => {
  if (condition) {
    return obj
  }
  if (isArray) {
    return []
  }
  return {}
}

// https://github.com/styled-components/styled-components/issues/630
export function withProps<U>() {
  return <
    P extends keyof JSX.IntrinsicElements | React.ComponentType<any>,
    T extends object,
    O extends object = {}
  >(
    fn: ThemedStyledFunction<P, T, O>
  ): ThemedStyledFunction<P & U, T, O & U> =>
    (fn as unknown) as ThemedStyledFunction<P & U, T, O & U>
}

// Color utils

/**
 * Checks if the color is light or darks based on its L value in
 * the HSL color scale.
 * @param color A chroma color instance
 */
export const isLightColor = color => color.get("hsl.l") >= 0.5

/**
 * Returns font color of updated lightness on hover
 * @param color A chroma color instance
 */
export const getHoverFontColor = color => {
  return isLightColor(color)
    ? color.set("hsl.l", "-0.3")
    : color.set("hsl.l", "+0.3")
}

/**
 * Returns background color of updated lightness on hover
 * @param color A chroma color instance
 */
export const getHoverBackgroundColor = color => {
  return isLightColor(color)
    ? color.set("hsl.l", "-0.1")
    : color.set("hsl.l", "+0.1")
}


/**
 * Determines the appropriate font color based on the background color.
 */
export const getFontColor = (theme, backGroundColor) => {
  return isLightColor(backGroundColor)
    ? theme.chromaPalette.Neutral90
    : theme.chromaPalette.Neutral0
}

const createParsedColorTheme = (theme, backgroundColor) => ({
  background: backgroundColor,
  font: getFontColor(theme, backgroundColor),
})

/**
 * Creates an object with background and font properties initialized
 * with appropriate chroma colors.
 * @param backGroundColor A plain CSS color string
 */
export const parseCustomColor = (theme, customColor: CustomColor) => {
  if (
    !(
      chroma.valid(customColor) ||
      (typeof customColor === "object" &&
        chroma.valid(customColor.color) &&
        chroma.valid(customColor.secondaryColor))
    )
  ) {
    throw new Error("Provided color theme contains invalid color values")
  }
  if (typeof customColor === "object") {
    return {
      font: chroma(customColor.secondaryColor),
      background: chroma(customColor.color),
    }
  }
  customColor = chroma(customColor)
  return createParsedColorTheme(theme, customColor)
}

/**
 * Creates an object with background and font properties initialized
 * with appropriate chroma colors.
 * @param backGroundColor One of the predefined preset values
 */
export const parseColorPreset = (theme, backgroundColor: ColorPreset) => {
  backgroundColor = theme.secondaryPalette[backgroundColor]
  return createParsedColorTheme(theme, backgroundColor)
}

/**
 * Get Color from Styled Component Props
 */
export const getThemeColor = (props: any, color: string) => {
  const {
    theme: {
      knitui: { chromaPalette },
    },
  } = props

  return chromaPalette[color].hex()
}

/**
 * Get OS of current platform
 */
export const getOSName = () => {
  const platform = window.navigator.platform
  const userAgent = window.navigator.userAgent

  const macOSList = ["Macintosh", "MacIntel", "MacPPC", "Mac68K"]
  const windowOSList = ["Windows", "Win16", "Win32", "WinCE"]
  const iOSList = ["iPhone", "iPad", "iPod"]

  let os: string | undefined

  if (macOSList.includes(platform)) {
    os = "MacOS"
  } else if (windowOSList.includes(platform)) {
    os = "Windows"
  } else if (iOSList.includes(platform)) {
    os = "iOS"
  } else if (/Android/.test(userAgent)) {
    os = "Android"
  } else if (/Linux/.test(userAgent)) {
    os = "Linux"
  }

  return os
}
