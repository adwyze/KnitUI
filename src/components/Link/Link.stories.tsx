import React from "react"
import { storiesOf } from "@storybook/react"
import { withKnobs } from "@storybook/addon-knobs"
import Link from "./"

const stories = storiesOf("Link", module)
stories.addDecorator(withKnobs)

stories
  .addParameters({
    readme: {
      // Show readme before story
      codeTheme: "shades-of-purple",
      // Show readme at the addons panel
    },
  })
  .add("Basic", () => (
    <Link href="https://github.com/clarisights/KnitUI">KnitUI</Link>
  ))
  .add("Without underline", () => (
    <Link underline={false} href="https://github.com/clarisights/KnitUI">
      KnitUI
    </Link>
  ))
  .add("With newTab", () => (
    <Link underline newTab href="https://github.com/clarisights/KnitUI">
      KnitUI
    </Link>
  ))
  .add("With disabled and className", () => (
    <Link
      underline
      disabled
      newTab
      className="knit-link"
      href="https://github.com/clarisights/KnitUI">
      KnitUI
    </Link>
  ))
