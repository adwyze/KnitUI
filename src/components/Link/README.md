# Link

## Usage

```javascript
import { Link } from "KnitUI"

return <Link href="...">Link Text</Link>
```

| Prop name | Type       | Optional | Default |
| --------- | ---------- | -------- | ------- |
| disabled  | boolean    | Yes      | true    |
| className | string     | Yes      | None    |
| style     | CSS Object | Yes      | None    |
| newTab    | boolean    | Yes      | false   |
| underline | boolean    | Yes      | true    |
