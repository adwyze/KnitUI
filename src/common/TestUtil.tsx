import React, { ReactElement } from "react"
import { render } from "react-testing-library"
import { ThemeProvider } from "./styles"

const ProviderWrapper = ({
  children,
}: {
  children: ReactElement
}): ReactElement => {
  // Nest all providers here
  return <ThemeProvider>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: object) =>
  render(ui, { wrapper: ProviderWrapper, ...options })

// re-export everything
export * from "react-testing-library"
// override render method
export { customRender as render }
