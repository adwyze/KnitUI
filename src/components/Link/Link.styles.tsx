import styled from "styled-components"
import { getThemeColor, getHoverBackgroundColor as getHoverColor } from "../../common/_utils"
import { ILinkProps } from "./types"
import chroma from "chroma-js"

export const StyledLink = styled.a<ILinkProps>`
  text-decoration: none;
  padding: 0.4rem 0.2rem;
  font-style: normal;
  font-weight: 500;
  font-size: 1.4rem;
  line-height: 2rem;
  outline: none;
  color: ${props => getThemeColor(props, "Blue100")};
  text-decoration-line: ${props => (props.underline ? "underline" : "none")};
  border-radius: 0.4rem;
  border: 0.1rem solid transparent;
  text-underline-position: under;
  &[disabled] {
    cursor: not-allowed;
    opacity: 0.6;
    &:active {
      pointer-events: none;
    }
  }
  &:not([disabled]) {
    &:focus {
      box-sizing: border-box;
      border: 0.1rem solid ${props => getThemeColor(props, "Azure80")};
      box-shadow: 0rem 0rem 0.2rem #0066ff;
    }
    &:hover {
      color: ${props => getHoverColor(chroma(getThemeColor(props, "Blue100")))};
    }
  }
`
