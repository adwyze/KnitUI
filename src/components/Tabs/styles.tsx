import { RefObject } from "react"
import styled, { createGlobalStyle } from "styled-components"
import { Button, Icon } from ".."
import { getThemeColor, getOSName } from "../../common/_utils"

/* The tab being dragged is appended to the body, hence it's styles need to be at global scope*/

export const GlobalStyle = createGlobalStyle`
.knitui-tabs-helper:not(#knit-active-tab) {
  background: ${props => getThemeColor(props, "Beige10")};
  
  &:before {
    content: "";
    position: absolute;
    left: 0;
    top: 50%;
    transform: translateY(-50%);
    height: 1.4rem;
    width: 1px;
    background: ${props => getThemeColor(props, "Neutral40")};
  }
}

.knitui-tabs-helper {
  white-space: nowrap;
}
`

export const ButtonWrapper = styled(Button)`
  border-radius: 4px 4px 0 0;
  :hover,
  :focus,
  :active {
    background: #efe9dc;
  }
`

export const TabsWrapper = styled.div<{ hideTabContent: boolean }>`
  height: ${({ hideTabContent }) => (hideTabContent ? "auto" : "100%")};
  width: 100%;
`

export const TabsPanelWrapper = styled.div`
  display: flex;
  background: ${props => getThemeColor(props, "Beige10")};
`

const getActiveTabHeight = props => {
  return "100%"

  /**
   * Idea was to assign height as much as active tab but during drag tab
   * height has extra 2px which causing recursive increase of OverflowWrapper's height.
   * Have to find work-around for this.
   */
  return props.activeTabRef.current
    ? `${props.activeTabRef.current.getBoundingClientRect().height}px`
    : "100%"
}

export const OverflowWrapper = styled.div<{
  activeTabRef: RefObject<HTMLDivElement>
}>`
  /**
   * Height needed to taken from ref because in some cases like in MacOS Mojave 
   * irrespective of browser & it's version, padding-bottom still visible.
   */
  height: ${props => getActiveTabHeight(props)};
  overflow: hidden;
  position: relative;
  display: flex;
`

export const BlurElement = styled.div<{ dir: string; visible: boolean }>`
  width: 80px;
  height: 100%;
  position: absolute;
  background: linear-gradient(
    to ${props => (props.dir === "left" ? "right" : "left")},
    ${props => getThemeColor(props, "Beige10")}FF,
    ${props => getThemeColor(props, "Beige10")}00
  );
  z-index: 100;
  visibility: ${props => (props.visible ? "visible" : "hidden")};
  pointer-events: none;
`

export const IconWrapper = styled.div<{ visible: boolean }>`
  display: flex;
  justify-content: center;
  align-items: center;
  transition: visibility 250ms ease, width 150ms linear;
  visibility: ${props => (props.visible ? "visible" : "hidden")};
  width: ${props => (props.visible ? "30px" : "0")};
  cursor: pointer;
  z-index: 100;

  :hover,
  :focus,
  :active {
    background: #efe9dc;
  }
`

export const TabPanel = styled.button<{
  active?: boolean
  dragHandle?: boolean
}>`
  display: flex;
  flex-wrap: nowrap;
  justify-content: center;
  align-items: center;
  word-wrap: none;

  background: ${props =>
    props.active ? getThemeColor(props, "Neutral0") : "none"};
  color: ${props =>
    props.active
      ? getThemeColor(props, "Neutral90")
      : getThemeColor(props, "Neutral50")};
  border: 2px solid transparent;
  border-top: ${props =>
    props.active ? "2px solid #D8C9A7" : "2px solid transparent"};
  padding: 4px 14px;
  border-radius: 4px 4px 0px 0px;
  min-width: 80px;
  line-height: 20px;
  font-size: 14px;
  cursor: ${props => (props.dragHandle && props.active ? "auto" : "pointer")};

  /* To prevent focus outline overlapping with divider */
  margin-right: ${props => (props.active ? "0" : "1px")};

  :hover {
    background: ${props => (props.active ? "#FFFFFF" : "#efe9dc")};
  }

  :active,
  :focus {
    z-index: 101 !important;
    outline-color: ${props => getThemeColor(props, "Azure80")};
    border-color: ${props => getThemeColor(props, "Azure80")} auto 1px;
  }
`

export const TabContentWrapper = styled.div`
  width: 100%;
`

export const OverflowContainer = styled.div`
  display: flex;
  align-items: center;
  white-space: nowrap;
  width: 100%;
  overflow: scroll hidden;
  padding-bottom: ${getOSName() !== "MacOS" ? "0px" : "17px"};
  margin-bottom: -17px;
  box-sizing: content-box;
`

export const DragIcon = styled(Icon)`
  cursor: grab;
  & svg {
    fill: ${props => getThemeColor(props, "Neutral45")};
  }
  margin-right: 0.4rem;
  & svg :hover,
  & svg :active,
  & svg :focus {
    fill: #000000;
  }
`

export const VerticalBar = styled.div`
  height: 14px;
  width: 0px;
  position: absolute;
  border-right: 1px solid #cccccc;
  top: 50%;
  right: 0;
  transform: translateY(-50%);
  pointer-events: none;
`
