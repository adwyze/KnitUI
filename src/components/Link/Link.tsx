import React from "react"
import { ILinkProps } from "./types"
import { StyledLink } from "./Link.styles"

const Link = ({
  className,
  style,
  underline,
  href,
  children,
  disabled,
  newTab,
}: ILinkProps) => {
  const newTabProps = {
    target: "_blank",
    rel: "noopener noreferrer",
  }
  return (
    <StyledLink
      underline={underline}
      className={className}
      style={style}
      href={href}
      disabled={disabled}
      {...(newTab && newTabProps)}>
      {children}
    </StyledLink>
  )
}

Link.defaultProps = {
  underline: true,
  disabled: false,
  newTab: false,
}

export default Link
