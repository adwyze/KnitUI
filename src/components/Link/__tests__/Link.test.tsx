import React from "react"
import Link from "../Link"
import { render, cleanup, fireEvent } from "../../../common/TestUtil"
import { ThemeProvider } from "../../../common/styles"

afterEach(cleanup)

describe("Link component", () => {
  it("should render correctly", () => {
    const { asFragment } = render(
      <ThemeProvider>
        <Link href="https://github.com/clarisights/KnitUI">KnitUI</Link>
      </ThemeProvider>
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("should not work when disabled", () => {
    const { asFragment } = render(
      <ThemeProvider>
        <Link
          className="knit-link"
          disabled
          href="https://github.com/clarisights/KnitUI">
          KnitUI
        </Link>
      </ThemeProvider>
    )

    const element = document.querySelector(".knit-link")

    expect(element).toHaveAttribute("disabled")
    expect(element).toHaveStyle("cursor: not-allowed")
    expect(element).toHaveStyle("opacity: 0.6")

    expect(asFragment()).toMatchSnapshot()
  })
  it("should support new tab", () => {
    const { asFragment } = render(
      <ThemeProvider>
        <Link
          className="knit-link"
          newTab
          href="https://github.com/clarisights/KnitUI">
          KnitUI
        </Link>
      </ThemeProvider>
    )

    const element = document.querySelector(".knit-link")

    expect(element).toHaveAttribute("target", "_blank")
    expect(element).toHaveAttribute("rel", "noopener noreferrer")

    expect(asFragment()).toMatchSnapshot()
  })
  it("should support underline prop", () => {
    const { asFragment } = render(
      <ThemeProvider>
        <Link
          className="knit-link"
          underline={false}
          href="https://github.com/clarisights/KnitUI">
          KnitUI
        </Link>
      </ThemeProvider>
    )

    const element = document.querySelector(".knit-link")

    expect(element).toHaveStyle("text-decoration-line: none")

    expect(asFragment()).toMatchSnapshot()
  })
})
