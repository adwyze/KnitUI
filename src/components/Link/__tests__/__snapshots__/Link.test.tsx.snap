// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component should not work when disabled 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 0.4rem 0.2rem;
  font-style: normal;
  font-weight: 500;
  font-size: 1.4rem;
  line-height: 2rem;
  outline: none;
  color: #002966;
  -webkit-text-decoration-line: underline;
  text-decoration-line: underline;
  border-radius: 0.4rem;
  border: 0.1rem solid transparent;
  text-underline-position: under;
}

.c0[disabled] {
  cursor: not-allowed;
  opacity: 0.6;
}

.c0[disabled]:active {
  pointer-events: none;
}

.c0:not([disabled]):focus {
  box-sizing: border-box;
  border: 0.1rem solid #5ab2f2;
  box-shadow: 0rem 0rem 0.2rem #0066ff;
}

.c0:not([disabled]):hover {
  color: #003d99;
}

<a
    class="knit-link c0"
    disabled=""
  >
    KnitUI
  </a>
</DocumentFragment>
`;

exports[`Link component should render correctly 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 0.4rem 0.2rem;
  font-style: normal;
  font-weight: 500;
  font-size: 1.4rem;
  line-height: 2rem;
  outline: none;
  color: #002966;
  -webkit-text-decoration-line: underline;
  text-decoration-line: underline;
  border-radius: 0.4rem;
  border: 0.1rem solid transparent;
  text-underline-position: under;
}

.c0[disabled] {
  cursor: not-allowed;
  opacity: 0.6;
}

.c0[disabled]:active {
  pointer-events: none;
}

.c0:not([disabled]):focus {
  box-sizing: border-box;
  border: 0.1rem solid #5ab2f2;
  box-shadow: 0rem 0rem 0.2rem #0066ff;
}

.c0:not([disabled]):hover {
  color: #003d99;
}

<a
    class="c0"
    href="https://github.com/clarisights/KnitUI"
  >
    KnitUI
  </a>
</DocumentFragment>
`;

exports[`Link component should support new tab 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 0.4rem 0.2rem;
  font-style: normal;
  font-weight: 500;
  font-size: 1.4rem;
  line-height: 2rem;
  outline: none;
  color: #002966;
  -webkit-text-decoration-line: underline;
  text-decoration-line: underline;
  border-radius: 0.4rem;
  border: 0.1rem solid transparent;
  text-underline-position: under;
}

.c0[disabled] {
  cursor: not-allowed;
  opacity: 0.6;
}

.c0[disabled]:active {
  pointer-events: none;
}

.c0:not([disabled]):focus {
  box-sizing: border-box;
  border: 0.1rem solid #5ab2f2;
  box-shadow: 0rem 0rem 0.2rem #0066ff;
}

.c0:not([disabled]):hover {
  color: #003d99;
}

<a
    class="knit-link c0"
    href="https://github.com/clarisights/KnitUI"
    rel="noopener noreferrer"
    target="_blank"
  >
    KnitUI
  </a>
</DocumentFragment>
`;

exports[`Link component should support underline prop 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 0.4rem 0.2rem;
  font-style: normal;
  font-weight: 500;
  font-size: 1.4rem;
  line-height: 2rem;
  outline: none;
  color: #002966;
  -webkit-text-decoration-line: none;
  text-decoration-line: none;
  border-radius: 0.4rem;
  border: 0.1rem solid transparent;
  text-underline-position: under;
}

.c0[disabled] {
  cursor: not-allowed;
  opacity: 0.6;
}

.c0[disabled]:active {
  pointer-events: none;
}

.c0:not([disabled]):focus {
  box-sizing: border-box;
  border: 0.1rem solid #5ab2f2;
  box-shadow: 0rem 0rem 0.2rem #0066ff;
}

.c0:not([disabled]):hover {
  color: #003d99;
}

<a
    class="knit-link c0"
    href="https://github.com/clarisights/KnitUI"
  >
    KnitUI
  </a>
</DocumentFragment>
`;
